lib/strategy/FragmentationStrategyShape.ts

```diff
@@ -0,0 +1,292 @@
+   * @param {string} iri - the targeted iri where the quad will be pushed
+   */
+  public static async generateShapeTreeLocator(quadSink: IQuadSink,
+    podIRI: string,
+    shapeTreeIRI: string,
+    iri: string): Promise<void> {
+    const shapeTreeIndicator = DF.quad(
+      DF.namedNode(podIRI),
+      this.shapeTreeLocator,
+      DF.namedNode(shapeTreeIRI),
+    );
+    await quadSink.push(iri, shapeTreeIndicator);
+  }
+
+  /**
+   * Generate the triples of a shapetree file that target the content and a shape
+   * in a pod and push them into a quad sink .
+   * The triples generated are not compliante with the shapetree specification because we are in the process
+   * of creating a shape index method inspired from the type index and shapetree specification.
+   * @param {IQuadSink} quadSink - a quad sink
+   * @param {string} shapeTreeIRI - the iri of the shapetree
+   * @param {string} shapeIRI - the iri of the target shape
+   * @param {boolean} isInRootOfPod - indicate if the resource is at the root of the pod
+   * @param {string} contentIri - the iri of the target resource
+   */
+  public static async generateShapetreeTriples(quadSink: IQuadSink,
+    shapeTreeIRI: string,
+    shapeIRI: string,
+    isInRootOfPod: boolean,
+    contentIri: string): Promise<void> {
+    const blankNode = DF.blankNode();
+    const shape = DF.quad(
+      blankNode,
+      this.shapeTreeShapeNode,
+      DF.namedNode(shapeIRI),
+    );
+    // We need to indicate if the shape index target a single file or if it is a container that is targeted
+    const target = DF.quad(
+      blankNode,
+      isInRootOfPod ? this.solidInstance : this.solidInstanceContainer,
+      DF.namedNode(contentIri),
+    );
+    await Promise.all(
+      [
+        quadSink.push(shapeTreeIRI, shape),
+        quadSink.push(shapeTreeIRI, target),
+      ],
+    );
+  }
+
+  /**
+   * Generate an ShExJ shape from a SheXC file and push the quads into a quad sink
+   * @param {IQuadSink} quadSink - a quad sink
+   * @param {string} shapeIRI - The iri of the target shape
+   * @param {string} shapeShexc - ShExC string
+   */
+  public static async generateShape(quadSink: IQuadSink, shapeIRI: string, shapeShexc: string): Promise<void> {
+    const shexParser = ShexParser.construct(shapeIRI);
+    const shapeJSONLD = shexParser.parse(shapeShexc);
+    const stringShapeJsonLD = JSON.stringify(shapeJSONLD);
+
+    return new Promise((resolve, reject) => {
+      const promises: Promise<void>[] = [];
+      // The jsonLD is not valid without the context field and the library doesn't include it
+      // because a ShExJ MAY contain a @context field
+      // https://shex.io/shex-semantics/#shexj
+      const jsonldParser = new JsonLdParser({
+        streamingProfile: false,
+        context: SHEX_CONTEXT,
+        skipContextValidation: true,
+      });
+      jsonldParser
+        .on('data', async(quad: RDF.Quad) => {
+          promises.push(quadSink.push(shapeIRI, quad));
```

Comment:
The danger here is that promise rejections may go uncaught if the parser is slow, and Promise.all isn't started immediately.

I recommend just collecting the quads in a sync-manner, and once the parser ends, pushing them all to the sink in a for-loop with await on each separate promise.